File: Baekjoon_Solve/Recursion/20008/20008.java
import java.io.*;
import java.util.StringTokenizer;


public class Main {

    static class Skill {
        int c;
        int d;
        Skill(int c, int d) {
            this.c = c;
            this.d = d;
        }
    }

    static void backtrack(int left, int t) {
        if (left <= 0) {
            ans = Math.min(ans, t);
            return;
        }
        boolean flag = false; // 쿨타임이 차서 스킬을 사용할 수 있음 유무
        for (int i = 0; i < N; i++) {
            Skill skill = skills[i];
            if (time[i] + skill.c <= t) { // 스킬 i를 사용한 시간에 쿨을 더 했을 때의 시간이 t보다 작거나 같으면 사용 가능
                int tmp = time[i];
                time[i] = t; // i 번째 스킬을 사용한 시간을 기록
                backtrack(left - skill.d, t + 1);
                time[i] = tmp;
                flag = true;
            }
        }
        if (!flag) backtrack(left, t + 1);
    }

    static int N, HP;
    static Skill[] skills;
    static int[] time; // i번 스킬을 사용한 시간
    static int ans = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        HP = Integer.parseInt(st.nextToken());

        time = new int[N];
        skills = new Skill[N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int c = Integer.parseInt(st.nextToken());
            int d = Integer.parseInt(st.nextToken());
            skills[i] = new Skill(c, d);
            time[i] -= c; // 0초에 스킬을 사용하기 위해 음수로 기록
        }
        backtrack(HP, 0);
        
        System.out.println(ans);
    }
}

File: Baekjoon_Solve/Recursion/20008/20008.py
# 몬스터를 처치하자! (20008번)
import sys
input = sys.stdin.readline

N, HP = map(int, input().rstrip().split())
skills = []
for i in range(N):
    skills.append(tuple(map(int, input().rstrip().split())))

cooltime = [0 for _ in range(N)]  # 쿨타임 배열
ans = int(1e9)


def backtrack(cur: int, left: int) -> None:
    global ans
    if left <= 0:
        ans = min(ans, cur)
        return
    flag = False  # 현재 시간에 스킬을 사용 가능
    for i in range(N):
        t = skills[i][0]  # 딜레이
        d = skills[i][1]  # 데미지
        if cooltime[i] <= cur:  # 쿨타임이 다 찬 시간보다 크면 스킬을 사용 가능
            flag = True
            last = cooltime[i]
            cooltime[i] = cur + t  # 다음 사용 가능한 시간 기록
            backtrack(cur + 1, left - d)
            cooltime[i] = last

    if not flag:  # 스킬 사용이 불가능하면 시간을 1초 증가시켜서 탐색
        backtrack(cur + 1, left)


backtrack(0, HP)
print(ans)
